etc/init.d/common.init.in

```diff
@@ -0,0 +1,405 @@
+#   2 if daemon could not be started
+#   3 if unsupported
+#
+zfs_daemon_start()
+{
+	local PIDFILE="$1";	shift
+	local DAEMON_BIN="$1";	shift
+	local DAEMON_ARGS="$*"
+
+	if type start-stop-daemon > /dev/null 2>&1 ; then
+		# LSB functions
+		start-stop-daemon --start --quiet --pidfile "$PIDFILE" \
+		    --exec "$DAEMON_BIN" --test > /dev/null || return 1
+
+	        start-stop-daemon --start --quiet --exec "$DAEMON_BIN" -- \
+		    $DAEMON_ARGS || return 2
+	elif type daemon > /dev/null 2>&1 ; then
+	        # Fedora/RedHat functions
+		daemon --pidfile "$ZED_PIDFILE" "$ZED" $DAEMON_ARGS
+		return $?
+	else
+		# Unsupported
+		return 3
+	fi
+
+	return 0
+}
+
+# Returns
+#   0 if daemon has been stopped
+#   1 if daemon was already stopped
+#   2 if daemon could not be stopped
+#   3 if unsupported
+#
+zfs_daemon_stop()
+{
+	local PIDFILE="$1"
+	local DAEMON_BIN="$2"
+	local DAEMON_NAME="$3"
+
+	if type start-stop-daemon > /dev/null 2>&1 ; then
+		# LSB functions
+		start-stop-daemon --stop --quiet --retry=TERM/30/KILL/5 \
+		    --pidfile "$PIDFILE" --name "$DAEMON_NAME"
+		return $?
+	elif type killproc > /dev/null 2>&1 ; then
+		# Fedora/RedHat functions
+		killproc "$DAEMON_NAME"
+		return $?
+	else
+		# Unsupported
+		return 3
+	fi
+
+	return 0
+}
+
+# Returns status
+zfs_daemon_status()
+{
+	local DAEMON_BIN="$1"
+	local DAEMON_NAME="$2"
+
+	if type status_of_proc > /dev/null 2>&1 ; then
+		# LSB functions
+		status_of_proc "$DAEMON_NAME" "$DAEMON_BIN"
+		return $?
+	elif type status > /dev/null 2>&1 ; then
+		# Fedora/RedHat functions
+		status "$DAEMON_NAME"
+		return $?
+	else
+		# Unsupported
+		return 3
+	fi
+
+	return 0
+}
+
+zfs_daemon_reload()
+{
+	local PIDFILE="$1"
+	local DAEMON_NAME="$2"
+
+	if type start-stop-daemon > /dev/null 2>&1 ; then
+		# LSB functions
+		start-stop-daemon --stop -signal 1 --quiet \
+		    --pidfile "$PIDFILE" --name "$DAEMON_NAME"
+		return $?
+	elif type killproc > /dev/null 2>&1 ; then
+		# Fedora/RedHat functions
+                killproc "$DAEMON_NAME" "-SIGHUP" > /dev/null
+		return $?
+	else
+		# Unsupported
+		return 3
+	fi
+
+	return 0
+}
+
+zfs_installed()
+{
+	if [ ! -x "$ZPOOL" ]; then
+		return 1
+	else
+		# Test if it works (will catch missing/broken libs etc)
+		"$ZPOOL" -? > /dev/null 2>&1 
+		return $?
+	fi
+
+	if [ ! -x "$ZFS" ]; then
+		return 2
+	else
+		# Test if it works (will catch missing/broken libs etc)
+		"$ZFS" -? > /dev/null 2>&1
+		return $?
+	fi
+
+	return 0
+}
+
+# Trigger udev and wait for it to settle.
+udev_trigger()
+{
+	if [ -x /sbin/udevadm ]; then
+		/sbin/udevadm trigger --action=change --subsystem-match=block
+		/sbin/udevadm settle
+	elif [ -x /sbin/udevsettle ]; then
+		/sbin/udevtrigger
+		/sbin/udevsettle
+	fi
+}
+
+# From scripts/common.sh
+wait_udev()
+{
+	local DEVICE=$1
+	local DELAY=$2
+	local COUNT=0
+
+	udev_trigger
+	while [ ! -e "${DEVICE}" ]; do
+		if [ ${COUNT} -gt "${DELAY}" ]; then
+			return 1
+		fi
+
+		let COUNT=${COUNT}+1
+		sleep 1
```

Comment:
@ryao Assuming that it will ONLY be run at the boot procedure is wrong, It should be able to run at any time (with or without busybox etc). And if you checked my test case, it doesn't even care if the module is loaded or anything. It was just a [simple] fork bomb.

Also, although the scheduler is the same, the resources is not. Running <code>echo</code> fifteen hundred times WILL use unnecessary resources. I also think that "The Gentoo" developers (including you and everyone else that think using busy-loops is a good idea in shell scripts etc) is dead wrong. I think one [of many] reasons why Linus decided to do his own version of minix, was that it didn't have a proper scheduler. And keeping that occupied with running <code>echo</code> (or whatever) is just wrong! NASA couldn't use msdos for example for their satellites (etc) because it didn't have sleep (yes, really - a missing sleep! - it simply drained to much energy). So on system not powered by the wall socket, busy-loops is evil! Not that I can see that anyone would use ZoL/ZFS on such a system, but still :)

That said, I'm all for #3431, but could/should we assume that these script should ONLY be used with ZoL that have that feature? We probably could/should, but should we [assume that]?

Besides, #3431 is using sleep!! Micro sleep, granted, but that's just sleep that can wait a shorter amount of time [than whole seconds], so that's not really a busy-loop anyway.

But if you think waiting a whole second each time (which I think is silly to argue about really), I can wait 0.1 second etch time if that makes you feel better about it…

```
# time /bin/sleep 0.1

real    0m0.102s
user    0m0.000s
sys     0m0.000s
```